add support for zero-length boxes

chemag opened the pull request:
In order to end a list of sub-boxes, some ISOBMFF multiplexers prefer to end a box list with a length=0 entry, instead of a non-existent length field. Patch adds support for this behavior.

chemag committed:
isobmff: fix Makefile for Fedora

chemag committed:
add support for zero-length boxes

In order to end a list of sub-boxes, some ISOBMFF multiplexers prefer to
end a box list with a length=0 entry, instead of a non-existent length
field. Patch adds support for this behavior.

chemag commented:
ping

chemag committed:
add CTSS box

chemag commented:
added CTSS parser

chemag committed:
Add signed int reader

Only adding 32-bit BE, as it is the one needed for now.

chemag committed:
Use signed int32_t for CTTS offsets

Standard defines offsets (sample_offset) to be unsigned or signed
int depending on the version field. More concretely,
ISO/IEC 14496-12:2022 modifies the behavior of CTTS so that
makes sample_offset is a 32-bit signed int when version is 1.

Section 8.6.1.3.2:
```
aligned(8) class CompositionOffsetBox
  extends FullBox('ctts', version, 0) {
  unsigned int(32) entry_count;
  int i;
  if (version==0) {
    for (i=0; i < entry_count; i++) {
      unsigned int(32) sample_count;
      unsigned int(32) sample_offset;
    }
  }
  else if (version == 1) {
    for (i=0; i < entry_count; i++) {
      unsigned int(32) sample_count;
      signed int(32) sample_offset;
    }
  }
}
```

Apple MOV muxers are known to use negative CTTS offsets even
without setting the version field to 1. It typically makes no
sense to have a sensor offset larger than 0x7fff,ffff, so we
will force signed int always.

chemag commented:
did you manage to take a look at these 2 patches?

chemag committed:
add AV01 box

hwknsj commented:
yo the homie @chemag is trying to help y'all

@DDNA-BuildAgent help them out, robot

Commit:
add mp4a box

chemag committed:
fix sample rate in mp4a box

Commit:
fix mp4a parsing failure

ChenDu-Meta committed:
isobmff: add unit test for Parser

Add unit test for ISOBMFF Parser using static binary stream input. The
binary stream was dumped from a mp4 video file.

ChenDu-Meta committed:
isobmff: add fuzzing support for Parser

Add fuzzing support to ISOBMFF with libfuzzer.
Add fuzzing to Parser

ChenDu-Meta committed:
isobmff: add unit test to META atom parser

Add unit test to META atom parser

ChenDu-Meta committed:
isobmff: add fuzzing to META atom parser

Add fuzzing to META atom parser

ChenDu-Meta committed:
isobmff: add exception catch to Parser test/fuzzer

Add exception catch to Parser unit test and fuzzer to  prevent crashing during fuzzing

ChenDu-Meta committed:
isobmff: fix potential memory leak for wrong box size

If wrong box size is given the box->ReadData may fail, resulting memory
leak as content is nor properly deleted.
Fix it by try-catch to delete content if box->ReadData failed.

chemag committed:
isobmff: add Build/ to .gitignore

chemag committed:
isobmff: convert build system to cmake

Tested: linux
```
$ CC=gcc CXX=g++ cmake  ..
-- The CXX compiler identification is GNU 15.1.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.3s)
-- Generating done (0.0s)
-- Build files have been written to: isobmff/build
$ make -j
[  1%] Building CXX object CMakeFiles/ISOBMFF.dir/ISOBMFF/source/AV01.cpp.o
...
[ 98%] Building CXX object CMakeFiles/ISOBMFF.dir/ISOBMFF/source/URN.cpp.o
[100%] Linking CXX shared library libISOBMFF.so
[100%] Built target ISOBMFF
```

```
$ CC=clang CXX=clang++ cmake ..
-- The CXX compiler identification is Clang 20.1.3
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.4s)
-- Generating done (0.0s)
-- Build files have been written to: isobmff/build
$ make -j
[  3%] Building CXX object CMakeFiles/ISOBMFF.dir/ISOBMFF/source/AVC1.cpp.o
...
[100%] Linking CXX shared library libISOBMFF.so
[100%] Built target ISOBMFF
```

Tested: MacOS
```
% CC=gcc CXX=g++ cmake  ..
-- The CXX compiler identification is AppleClang 17.0.0.17000013
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.8s)
-- Generating done (0.0s)
-- Build files have been written to: isobmff/build

% make -j
[  3%] Building CXX object CMakeFiles/ISOBMFF.dir/ISOBMFF/source/AVC1.cpp.o
...
[ 98%] Building CXX object CMakeFiles/ISOBMFF.dir/ISOBMFF/source/Matrix.cpp.o
[100%] Linking CXX shared library libISOBMFF.dylib
[100%] Built target ISOBMFF
```

```
% CC=clang CXX=clang++ cmake ..

-- The CXX compiler identification is AppleClang 17.0.0.17000013
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done (0.7s)
-- Generating done (0.0s)
-- Build files have been written to: isobmff/build

% make -j
[  1%] Building CXX object CMakeFiles/ISOBMFF.dir/ISOBMFF/source/AV01.cpp.o
...
[ 98%] Building CXX object CMakeFiles/ISOBMFF.dir/ISOBMFF/source/HDLR.cpp.o
[100%] Linking CXX shared library libISOBMFF.dylib
[100%] Built target ISOBMFF
```

chemag committed:
isobmff: use std::abs()

abs() is defined only for int in <cstdlib>.

std::streamoff is typically a long or even long long, so using
abs(int) may lead to truncation, overflow, or undefined behavior
when dealing with large offsets.

chemag committed:
isobmff: add version infra

chemag committed:
isobmff: version 0.1

panayotovmarko committed:
isobmff: add fuzzing to FTYP atom parser

panayotovmarko committed:
isobmff: add unit test to FTYP atom parser

chemag committed:
isobmff: fix META test class name

panayotovmarko committed:
isobmff: add unit test to MVHD atom parser

panayotovmarko committed:
isobmff: add fuzzing to MVHD atom parser

chemag committed:
isobmff: fix minor nit in MVHD test

panayotovmarko committed:
isobmff: add unit test to HDLR atom parser

panayotovmarko committed:
isobmff: add fuzzing to HDLR atom parser

panayotovmarko committed:
isobmff: add unit test to TKHD atom parser

panayotovmarko committed:
isobmff: add fuzzing to TKHD atom parser

chemag committed:
isobmff: remove fuzz binaries

chemag committed:
isobmff: ignore fuzz binaries in git

chemag committed:
isobmff: add results from 10^6 fuzz pass

Tested:
```
$ cd fuzz
$ RUNS=1000000 make -j -Oline fuzz
...
```

There are 2x OOM that need to be fixed.

chemag committed:
isobmff: add symbols to fuzz builds

This is the wrong solution, but it should unblock us.

The right solution is to make the fuzz/ directory process part of the cmake
build.

Tested:
```
$ git clone
$ mkdir build
$ cd build
$ CC=clang CXX=clang++ cmake -DCMAKE_BUILD_TYPE=DEBUG -DCMAKE_C_FLAGS_DEBUG="-g -O0" -DCMAKE_CXX_FLAGS_DEBUG="-g -O0" ..  # h264nal with debug symbols and clang (for fuzzer)
$ make -j
$ cd ../fuzz
$ rm ../Build/fuzz/*fuzzer
$ make -j
$ export LD_LIBRARY_PATH=../build/:${LD_LIBRARY_PATH}
$ gdb --args ../Build/fuzz/Parser_fuzzer ../fuzz/corpus/Parser_fuzzer/oom-d073e78f82d0bbc1af5174905b40150f8de11914
(gdb) r
...
    #14 0x7ffff7a4980b in ISOBMFF::BinaryStream::Read(unsigned long) /tmp/isobmff/ISOBMFF/source/BinaryStream.cpp:73:32
    #15 0x7ffff7a53486 in ISOBMFF::ContainerBox::ReadData(ISOBMFF::Parser&, ISOBMFF::BinaryStream&) /tmp/isobmff/ISOBMFF/source/ContainerBox.cpp:132:60
    #16 0x7ffff7a8b086 in ISOBMFF::Parser::Parse(ISOBMFF::BinaryStream&) /tmp/isobmff/ISOBMFF/source/Parser.cpp:220:32
    #17 0x7ffff7a8ac75 in ISOBMFF::Parser::Parse(std::vector<unsigned char, std::allocator<unsigned char>> const&) /tmp/isobmff/ISOBMFF/source/Parser.cpp:181:15
...
```

And the OOM issue is in BinaryStream.cpp:73.

panayotovmarko committed:
isobmff: fix OOMs from fuzzing

Tested:

Before:
```
$ cd build/
$ make
$ cd ../fuzz/
$ rm ../Build/fuzz/*fuzzer
$ make -j
$ export LD_LIBRARY_PATH=../build/:${LD_LIBRARY_PATH}
$ ../Build/fuzz/Parser_fuzzer ../fuzz/corpus/Parser_fuzzer/oom-d073e78f82d0bbc1af5174905b40150f8de11914
$ ../Build/fuzz/Parser_fuzzer ../fuzz/corpus/Parser_fuzzer/oom-d073e78f82d0bbc1af5174905b40150f8de11914
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 1943964596
INFO: Loaded 1 modules   (236 inline 8-bit counters): 236 [0x620ba0, 0x620c8c),
INFO: Loaded 1 PC tables (236 PCs): 236 [0x5d7880,0x5d8740),
../Build/fuzz/Parser_fuzzer: Running 1 inputs 1 time(s) each.
Running: ../fuzz/corpus/Parser_fuzzer/oom-d073e78f82d0bbc1af5174905b40150f8de11914
==3169869== ERROR: libFuzzer: out-of-memory (malloc(3758096456))
...

SUMMARY: libFuzzer: out-of-memory
```

After:
```
$ cd ../build/
$ make
$ cd ../fuzz/
$ rm ../Build/fuzz/*fuzzer
$ make -j
$ export LD_LIBRARY_PATH=../build/:${LD_LIBRARY_PATH}
$ ../Build/fuzz/Parser_fuzzer ../fuzz/corpus/Parser_fuzzer/oom-d073e78f82d0bbc1af5174905b40150f8de11914
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 1774353212
INFO: Loaded 1 modules   (236 inline 8-bit counters): 236 [0x620ba0, 0x620c8c),
INFO: Loaded 1 PC tables (236 PCs): 236 [0x5d7880,0x5d8740),
../Build/fuzz/Parser_fuzzer: Running 1 inputs 1 time(s) each.
Running: ../fuzz/corpus/Parser_fuzzer/oom-d073e78f82d0bbc1af5174905b40150f8de11914
Caught exception: Insufficient data available for read
Executed ../fuzz/corpus/Parser_fuzzer/oom-d073e78f82d0bbc1af5174905b40150f8de11914 in 1 ms
***
*** NOTE: fuzzing was not performed, you have only
***       executed the target code on a fixed set of inputs.
***
```

chemag committed:
isobmff: add results from 10^6 fuzz pass

Tested:
```
$ cd fuzz
$ RUNS=1000000 make -j -Oline fuzz
...
```

No OOM left.

panayotovmarko committed:
isobmff: fix build system

chemag committed:
isobmff: test new build system

chemag committed:
isobmff: cleanup old directory structure

chemag committed:
isobmff: version 0.2

panayotovmarko committed:
isobmff: fix build in XCode

panayotovmarko committed:
isobmff: add unit test to MDHD

panayotovmarko committed:
isobmff: add fuzzing to MDHD

chemag committed:
isobmff: add results from 10^7 fuzz pass

Tested:
```
$ cd fuzz
$ RUNS=10000000 make -j -Oline fuzz
...
```

panayotovmarko committed:
isobmff: fix unit test build

Detect all tests inside test directory

panayotovmarko committed:
isobmff: fix fuzzer build

Detect all fuzzers inside fuzz directory

panayotovmarko committed:
isobmff: fix ReadData function in DREF.cpp

panayotovmarko committed:
isobmff: add unit test to DREF atom

panayotovmarko committed:
isobmff: add fuzzing to DREF atom

chemag committed:
isobmff: add parameters to main.cpp

Tested:
```
$ tools/isobmff --no-analyze-flag in.mp4
$ tools/isobmff --analyze-flag in.mp4
[ ISOBMFF::File ]
{
    [ ftyp ]
    {
        - Major brand:       isom
...
$ tools/isobmff in.mp4
[ ISOBMFF::File ]
{
    [ ftyp ]
    {
        - Major brand:       isom
...
```

chemag committed:
isobmff: make skipping mdat copy the default

Tested:

default case:
```
$ valgrind --tool=massif --detailed-freq=1 ./tools/isobmff in.mp4
$ grep mem_heap_B massif.out.default | awk -F= '{if($2 > max) max=$2} END{print max}'
691167
```

donotskipmdat:

Added the following patch
```
diff --git a/tools/main.cpp b/tools/main.cpp
index def1861..5552cb7 100644
--- a/tools/main.cpp
+++ b/tools/main.cpp
@@ -223,6 +223,7 @@ int main( int argc, char *const * argv )

         try
         {
+            parser.AddOption( ISOBMFF::Parser::Options::DoNotSkipMDATData );
             parser.Parse( path );
         }
         catch( const std::runtime_error & e )
```

And then:
```
$ valgrind --tool=massif --detailed-freq=1 ./tools/isobmff in.mp4
$ grep mem_heap_B massif.out.donotskipmdat | awk -F= '{if($2 > max) max=$2} END{print max}'
682524151
```

chemag committed:
isobmff: make gcc happy

chemag committed:
isobmff: version 0.3

chemag committed:
isobmff: fix DREF unittest

Tested:
```
$ make test
Running tests...
/usr/local/lib64/python3.13/site-packages/cmake/data/bin/ctest
Test project isobmff/build
    Start 1: DREF_unittest
1/8 Test #1: DREF_unittest ....................   Passed    0.00 sec
    Start 2: FTYP_unittest
2/8 Test #2: FTYP_unittest ....................   Passed    0.00 sec
    Start 3: HDLR_unittest
3/8 Test #3: HDLR_unittest ....................   Passed    0.00 sec
    Start 4: MDHD_unittest
4/8 Test #4: MDHD_unittest ....................   Passed    0.00 sec
    Start 5: META_unittest
5/8 Test #5: META_unittest ....................   Passed    0.00 sec
    Start 6: MVHD_unittest
6/8 Test #6: MVHD_unittest ....................   Passed    0.00 sec
    Start 7: Parser_unittest
7/8 Test #7: Parser_unittest ..................   Passed    0.00 sec
    Start 8: TKHD_unittest
8/8 Test #8: TKHD_unittest ....................   Passed    0.00 sec

100% tests passed, 0 tests failed out of 8

Total Test time (real) =   0.03 sec
```

panayotovmarko committed:
isobmff: add unit test to STSD

panayotovmarko committed:
isobmff: add fuzzing to STSD

panayotovmarko committed:
isobmff: add unit test to HVC1 atom parser

panayotovmarko committed:
isobmff: add fuzzing to HVC1 atom parser

panayotovmarko committed:
isobmff: add unit test to HVCC atom parser

panayotovmarko committed:
isobmff: add fuzzing to HVCC atom parser

panayotovmarko committed:
isobmff: add unit test to STTS atom parser

panayotovmarko committed:
isobmff: add fuzzing to STTS atom parser

panayotovmarko committed:
isobmff: add unit test to STSS atom parser

panayotovmarko committed:
isobmff: add fuzzing to STSS atom parser

panayotovmarko committed:
isobmff: add unit test to MP4A atom parser

panayotovmarko committed:
isobmff: add fuzzing to MP4A atom parser

panayotovmarko committed:
isobmff: add unit test to URL atom parser

panayotovmarko committed:
isobmff: add fuzzing to URL atom parser

panayotovmarko committed:
isobmff: add unit test to CTTS atom parser

panayotovmarko committed:
isobmff: add fuzzing to CTTS atom parser

panayotovmarko committed:
isobmff: add results from 10^7 fuzz pass

Tested:
```
$ cd fuzz
$ RUNS=10000000 make -j -Oline fuzz
...
```

panayotovmarko committed:
isobmff: add unit test to AVC1 atom parser

panayotovmarko committed:
isobmff: add fuzzing to AVC1 atom parser

panayotovmarko committed:
isobmff: add uint test to AVCC atom parser

panayotovmarko committed:
isobmff: add fuzzing to AVCC atom parser

chemag committed:
isobmff: add version 0.4
